### PR TITLE
Add `npm_and_yarn` package managers' requirements to ecosystem metrics

### PR DESCRIPTION
**Pull request by kbukum1**
### What are you trying to accomplish?

This PR adds the ability to extract and report package manager version constraints for `npm`, `pnpm`, and `yarn` from the `engines` field in the `package.json` manifest file.

By including these constraints, the ecosystem metrics now provide better visibility into the range of supported package manager versions used in projects.

### What issues does this affect or fix?

This PR enhances ecosystem metrics tracking for the `npm_and_yarn` ecosystem by introducing package manager version requirement collection and normalization for `npm`, `pnpm`, and `yarn`.

### Anything you want to highlight for special attention from reviewers?

- Review the extraction logic for the `engines` field to ensure it handles constraints properly for `npm`, `pnpm`, and `yarn`.
- Validate that the conversion logic accurately transforms raw versions like `1.3.4` into `=1.3.4`.
- Consider edge cases where the `engines` field may contain incomplete or malformed version constraints.

### How will you know you've accomplished your goal?

- The extracted package manager constraints for `npm`, `pnpm`, and `yarn` are correctly represented in the ecosystem metrics.
- All tests pass, including new tests that cover the extraction and conversion logic.
- The system handles invalid or edge-case `engines` fields gracefully.

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb [hunk 1]
```diff
@@ -3,14 +3,45 @@
 
 require "dependabot/shared_helpers"
 require "dependabot/ecosystem"
+require "dependabot/npm_and_yarn/requirement"
 require "dependabot/npm_and_yarn/version_selector"
 
 module Dependabot
   module NpmAndYarn
-    ECOSYSTEM = "npm_and_yarn"
     MANIFEST_FILENAME = "package.json"
```
Comment (kbukum1) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:11]: Moved into global file

File: common/lib/dependabot/requirement.rb [hunk 2]
```diff
@@ -47,28 +49,89 @@ def max_version
...
+    end
+
+    sig { params(version: Dependabot::Version).returns(Dependabot::Version) }
+    def handle_tilde_pessimistic_min(version)
+      version
+    end
+
+    # Methods for handling maximum constraints
+    sig { params(version: Dependabot::Version).returns(Dependabot::Version) }
+    def handle_lte_max(version)
```
Comment (abdulapopoola) [inline common/lib/dependabot/requirement.rb:99]: can you write the method names in full?

Comment (kbukum1) [reply common/lib/dependabot/requirement.rb]: it is done.

**Review by abdulapopoola — APPROVED** [review]
